htdocs/compta/prelevement/class/bonprelevement.class.php

```diff
@@ -1043,33 +1072,20 @@ public function create($banque = 0, $agence = 0, $mode = 'real', $format = 'ALL'
 				$sql .= ", CONCAT(s.firstname,' ',s.lastname) as name";
 				$sql .= ", f.ref, sr.bic, sr.iban_prefix, 'FRST' as frstrecur";
 			}
-			if ($sourcetype != 'salary') {
-				if ($type != 'bank-transfer') {
-					$sql .= " FROM " . MAIN_DB_PREFIX . "facture as f";
-					$sql .= " LEFT JOIN " . MAIN_DB_PREFIX . "prelevement_demande as pd ON f.rowid = pd.fk_facture";
+			$sql .= " FROM " . $this->db->prefix() . $sqlTable . " as f";
+			$sql .= " LEFT JOIN " . $this->db->prefix() . "prelevement_demande as pd ON f.rowid = pd.fk_".$sqlTable;
+			$sql .= " LEFT JOIN " . $this->db->prefix() . $societeOrUser." as s ON s.rowid = f.".$socOrUser;
+			$sql .= " LEFT JOIN " . $this->db->prefix() . $societeOrUser."_rib as sr ON s.rowid = sr.".$socOrUser;
+			if ($sourcetype != 'salary'){
+				if (!empty($societeRibID)){
+					$sql .= " AND sr.rowid = " . $societeRibID;
 				} else {
-					$sql .= " FROM " . MAIN_DB_PREFIX . "facture_fourn as f";
-					$sql .= " LEFT JOIN " . MAIN_DB_PREFIX . "prelevement_demande as pd ON f.rowid = pd.fk_facture_fourn";
+					$sql .= " AND sr.default_rib = 1";
 				}
-				$sql .= " LEFT JOIN " . MAIN_DB_PREFIX . "societe as s ON s.rowid = f.fk_soc";
-				$sql .= " LEFT JOIN " . MAIN_DB_PREFIX . "societe_rib as sr ON s.rowid = sr.fk_soc AND sr.default_rib = 1";
-			} else {
-				$sql .= " FROM " . MAIN_DB_PREFIX . "salary as f";
-				$sql .= " LEFT JOIN " . MAIN_DB_PREFIX . "prelevement_demande as pd ON f.rowid = pd.fk_salary";
-				$sql .= " LEFT JOIN " . MAIN_DB_PREFIX . "user as s ON s.rowid = f.fk_user";
-				$sql .= " LEFT JOIN " . MAIN_DB_PREFIX . "user_rib as sr ON s.rowid = sr.fk_user";
-				// TODO Add 'AND sr.default_rib = 1' here. Note: the column has been created in v21 in llx_user_rib and default to 0
+				// TODO Add 'AND sr.default_rib = 1' in sourcetype salary too Note: the column has been created in v21 in llx_user_rib and default to 0
 				// If we add a test on sr.default_rib = 1, we must also check we have a correct error management to stop if no default BAN is found.
 			}
-			if ($sourcetype != 'salary') {
-				if ($type != 'bank-transfer') {
-					$sql .= " WHERE f.entity IN (" . getEntity('invoice') . ')';
-				} else {
-					$sql .= " WHERE f.entity IN (" . getEntity('supplier_invoice') . ')';
-				}
-			} else {
-				$sql .= " WHERE f.entity IN (" . getEntity('salary') . ')';
-			}
```

Comment:
This SQL request can still be improved upon when the user_rib case will be tackled. For now, llx_user HAS a default_rib column, but the work logic behind isn't here, so we still have the check on line 1079